### PR TITLE
feat: add minio log consumer with opts

### PR DESCRIPTION
**Pull request by nicufk**
## Checklist (choose whats happened)

- [ ] breaking change! (describe)
- [ ] tested locally
- [ ] tested on cluster
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

## Breaking changes

-

## Changes

-

## Fixes

-

### CODE REVIEW

File: pkg/logs/adapter/minio.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+		minioConnecter: minioconnecter.NewConnecter(endpoint, accessKeyID, secretAccessKey, region, token, bucket, log.DefaultLogger, opts...),
+		Log:            log.DefaultLogger,
+		bucket:         bucket,
+		region:         region,
+		disconnected:   false,
+		buffInfos:      make(map[string]BufferInfo),
+	}
+	minioClient, err := c.minioConnecter.GetClient()
+	if err != nil {
+		c.Log.Errorw("error connecting to minio", "err", err)
```
Comment (vsukhin) [inline pkg/logs/adapter/minio.go:50]: so, we ignore the errors, only logging them?

File: pkg/logs/adapter/minio.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	if _, ok := s.GetBuffInfo(id); !ok {
+		s.UpdateBuffInfo(id, BufferInfo{Buffer: bytes.NewBuffer(make([]byte, 0, defaultBufferSize)), Part: 0})
+	}
+
+	chunckToAdd, err := json.Marshal(e)
+	if err != nil {
+		return err
+	}
+	chunckToAdd = append(chunckToAdd, []byte("\n")...)
+	buffInfo, _ := s.GetBuffInfo(id)
```
Comment (vsukhin) [inline pkg/logs/adapter/minio.go:98]: so, we igore if it's ok or not ok?

File: pkg/logs/adapter/minio.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	writer := buffInfo.Buffer
+	_, err = writer.Write(chunckToAdd)
```
Comment (vsukhin) [inline pkg/logs/adapter/minio.go:100]: what will happen if len(chunckToAdd) > defaultBufferSize

File: pkg/logs/adapter/minio.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+		s.UpdateBuffInfo(id, buffInfo)
+		go s.putData(name, writer)
+	}
+
+	return nil
+}
+
+func (s *MinioConsumer) putData(name string, buffer *bytes.Buffer) {
+	if buffer != nil && buffer.Len() != 0 {
+		_, err := s.minioClient.PutObject(context.TODO(), s.bucket, name, buffer, int64(buffer.Len()), minio.PutObjectOptions{ContentType: "application/octet-stream"})
```
Comment (vsukhin) [inline pkg/logs/adapter/minio.go:118]: should we ue context as parameter?

File: pkg/logs/adapter/minio.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	if err != nil {
+		return err
+	}
+
+	if writer.Len() > defaultWriteSize {
+		buffInfo.Buffer = bytes.NewBuffer(make([]byte, 0, defaultBufferSize))
+		name := id + "-" + strconv.Itoa(buffInfo.Part)
+		buffInfo.Part++
+		s.UpdateBuffInfo(id, buffInfo)
+		go s.putData(name, writer)
```
Comment (vsukhin) [inline pkg/logs/adapter/minio.go:110]: sounds like making it async will always return sucess from this method

File: pkg/logs/adapter/minio.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+}
+
+func (s *MinioConsumer) combineData(minioClient *minio.Client, id string, parts int, deleteIntermediaryData bool) error {
+	buffer := bytes.NewBuffer(make([]byte, 0, parts*defaultBufferSize))
+	for i := 0; i < parts; i++ {
+		objInfo, err := minioClient.GetObject(context.TODO(), s.bucket, fmt.Sprintf("%s-%d", id, i), minio.GetObjectOptions{})
+		if err != nil {
+			s.Log.Errorw("error getting object", "err", err)
```
Comment (vsukhin) [inline pkg/logs/adapter/minio.go:133]: shouldn't we return err

File: pkg/logs/adapter/minio.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+		}
+		_, err = buffer.ReadFrom(objInfo)
+		if err != nil {
+			s.Log.Errorw("error reading object", "err", err)
```
Comment (vsukhin) [inline pkg/logs/adapter/minio.go:137]: shouldn't we return err?

File: pkg/logs/adapter/minio.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+	if err != nil {
+		s.Log.Errorw("error putting object", "err", err)
+		return err
+	}
+
+	if deleteIntermediaryData {
+		for i := 0; i < parts; i++ {
+			err = minioClient.RemoveObject(context.TODO(), s.bucket, fmt.Sprintf("%s-%d", id, i), minio.RemoveObjectOptions{})
+			if err != nil {
+				s.Log.Errorw("error removing object", "err", err)
```
Comment (vsukhin) [inline pkg/logs/adapter/minio.go:150]: shouldn't we return err

File: pkg/logs/adapter/minio.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+			}
+		}
+	}
+	buffer.Reset()
+	return nil
+}
+
+func (s *MinioConsumer) Stop(id string) error {
+	buffInfo, _ := s.GetBuffInfo(id)
```
Comment (vsukhin) [inline pkg/logs/adapter/minio.go:159]: should we ignore ok or not?

File: pkg/logs/adapter/minio_test.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+	consumer := NewMinioConsumer("localhost:9000", "minio", "minio123", "", "", bucket, minioconnecter.Insecure())
+	idChan := make(chan string, 100)
+	go verifyConsumer(idChan, bucket, consumer.minioClient)
+	var counter atomic.Int32
+	var wg sync.WaitGroup
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			//randomString := RandString(5)
```
Comment (vsukhin) [inline pkg/logs/adapter/minio_test.go:85]: do we need to have commented code?

File: pkg/logs/adapter/minio_test.go [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+			Type:    "test", Source: strconv.Itoa(i)})
+		totalSize += len(hugeString)
+		time.Sleep(time.Duration(rand.Intn(10)) * time.Millisecond)
+	}
+	sizeInMB := float64(totalSize) / 1024 / 1024
+	err := consumer.Stop(id)
+	if err != nil {
+		fmt.Println("#####error stopping", err)
+	}
+	//assert.NoError(b, err)
```
Comment (vsukhin) [inline pkg/logs/adapter/minio_test.go:111]: do we need to remove it?

File: pkg/logs/adapter/minio_test.go [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+	fmt.Printf("#####Total size for %s logs is %f MB\n\n\n", id, sizeInMB)
```
Comment (vsukhin) [inline pkg/logs/adapter/minio_test.go:112]: what is a purpose of printing in unit tests?

File: pkg/logs/adapter/minio_test.go [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+func verifyConsumer(idChan chan string, bucket string, minioClient *minio.Client) {
+	okSlice := make([]string, 0)
+	notOkSlice := make([]string, 0)
+	for id := range idChan {
+		//fmt.Println("#####verifying", id)
+		reader, err := minioClient.GetObject(context.Background(), bucket, id, minio.GetObjectOptions{})
+		if err != nil {
+			fmt.Println("######error getting object", err)
+		}
+		//assert.NoError(b, err)
```
Comment (vsukhin) [inline pkg/logs/adapter/minio_test.go:124]: should we remove it?

File: pkg/logs/adapter/minio_test.go [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+			//assert.Equal(b, strconv.Itoa(count), LogChunk.Source, "for id", id)
+			count++
+		}
+		if isOk {
+			okSlice = append(okSlice, id)
+		} else {
+			notOkSlice = append(notOkSlice, id)
+		}
+	}
+	fmt.Println("##### number of ok", len(okSlice))
```
Comment (vsukhin) [inline pkg/logs/adapter/minio_test.go:159]: what is a purpose of priniting in unit tests?

File: pkg/logs/adapter/minio_test.go [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+			//randomString := RandString(5)
+			randomString := strconv.Itoa(int(counter.Add(1)))
+			id := "test-bench" + "-" + randomString
+			testOneConsumer(consumer, id)
+			idChan <- id
+		}()
+	}
+	wg.Wait()
+	close(idChan)
+	fmt.Printf("#####Done buffInfo is %+v\n\n\n", consumer.buffInfos)
```
Comment (vsukhin) [inline pkg/logs/adapter/minio_test.go:193]: same

File: pkg/storage/minio/minio_connecter.go [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+			o.TlsConfig = &tls.Config{MinVersion: tls.VersionTLS12}
+		}
+		o.TlsConfig.Certificates = []tls.Certificate{cert}
+		o.Ssl = true
+		return nil
+	}
+}
+
+type Connecter struct {
+	Endpoint, AccessKeyID, SecretAccessKey, Region, Token, Bucket string
```
Comment (vsukhin) [inline pkg/storage/minio/minio_connecter.go:75]: might use separate line per field

**Review by vsukhin — COMMENTED** [review]
a few questions

File: pkg/logs/adapter/minio.go (hunk 5 above)
Comment (nicufk) [reply pkg/logs/adapter/minio.go:134]: The idea was to make Notify highly available and not to block for a lot of time, @exu what do you think?

File: pkg/logs/adapter/minio_test.go (hunk 12 above)
Comment (nicufk) [reply pkg/logs/adapter/minio_test.go:109]: it is a benchmark func that it is run in benchmarking conditions and it provides more information about it

**Review by vsukhin — APPROVED** [review]

**Pull request merged**